### PR TITLE
fix: mirror with local path

### PR DESCRIPTION
**Pull request by jiuker**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description

There have path with `/` start when local file path.
before:
`mc mirror --exclude "*.yaml" localPath` will not work.
now:
it work.

## Types of changes
- [x] Bug fix https://github.com/minio/mc/issues/4082
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fmc%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/difference_test.go [hunk 1]
```diff
@@ -24,26 +24,40 @@ import (
 var testCases = []struct {
 	pattern []string
 
-	object string
+	srcSuffix string
 
 	match bool
+
+	typ ClientURLType
```
Comment (shtripat) [inline cmd/difference_test.go:31]:
```suggestion
	type ClientURLType
```

Comment (klauspost) [reply cmd/difference_test.go:31]: @shtripat  `type` is a keywork.. Pretty sure you can't use it here.

Comment (shtripat) [reply cmd/difference_test.go:31]: ah yes. we are fine then

**Review by vadmeste — COMMENTED** [review]
@jiuker not sure, but would this work here, 'clnt.GetURL().Separator'  instead of passing the client type ? worth a try

**Comment by jiuker** [comment]
> @jiuker not sure, but would this work here, 'clnt.GetURL().Separator' instead of passing the client type ? worth a try

```go
func newClientURL(urlStr string) *ClientURL {
	scheme, rest := getScheme(urlStr)
	if strings.HasPrefix(rest, "//") {
		// if rest has '//' prefix, skip them
		var authority string
		authority, rest = splitSpecial(rest[2:], "/", false)
		if rest == "" {
			rest = "/"
		}
		host := getHost(authority)
		if host != "" && (scheme == "http" || scheme == "https") {
			return &ClientURL{
				Scheme:          scheme,
				Type:            objectStorage,
				Host:            host,
				Path:            rest,
				SchemeSeparator: "://",
				Separator:       '/',
			}
		}
	}
	return &ClientURL{
		Type:      fileSystem,
		Path:      rest,
		Separator: filepath.Separator,
	}
}
```
They can be the same.    @vadmeste

File: cmd/mirror-url.go [hunk 2]
```diff
@@ -89,7 +89,15 @@ func checkMirrorSyntax(ctx context.Context, cliCtx *cli.Context, encKeyDB map[st
...
 }
 
-func matchExcludeOptions(excludeOptions []string, srcSuffix string) bool {
+func matchExcludeOptions(excludeOptions []string, srcSuffix string, typ ClientURLType) bool {
+	// if type is file system, remove leading slash
+	if typ == fileSystem {
+		if strings.HasPrefix(srcSuffix, "/") {
+			srcSuffix = srcSuffix[1:]
+		} else if runtime.GOOS == "windows" && strings.HasPrefix(srcSuffix, "\\") {
+			srcSuffix = srcSuffix[2:]
```
Comment (klauspost) [inline cmd/mirror-url.go:98]: Why are you trimming 2 chars here? `"\\"` is just one char: `\`.

File: cmd/mirror-url.go [hunk 3]
```diff
@@ -89,7 +89,15 @@ func checkMirrorSyntax(ctx context.Context, cliCtx *cli.Context, encKeyDB map[st
 	return
 }
 
-func matchExcludeOptions(excludeOptions []string, srcSuffix string) bool {
+func matchExcludeOptions(excludeOptions []string, srcSuffix string, typ ClientURLType) bool {
+	// if type is file system, remove leading slash
+	if typ == fileSystem {
+		if strings.HasPrefix(srcSuffix, "/") {
+			srcSuffix = srcSuffix[1:]
+		} else if runtime.GOOS == "windows" && strings.HasPrefix(srcSuffix, "\\") {
```
Comment (harshavardhana) [inline cmd/mirror-url.go:97]:
```suggestion
		} else if runtime.GOOS == "windows" && strings.HasPrefix(srcSuffix, `\`) {
```

File: cmd/mirror-url.go (hunk 2 above)
Comment (harshavardhana) [inline cmd/mirror-url.go:98]:
```suggestion
			srcSuffix = srcSuffix[1:]
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Review by klauspost — APPROVED** [review]

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**